Userland/Libraries/LibGfx/ImageFormats/JPEG2000Loader.cpp

```diff
@@ -1119,11 +1202,165 @@ ErrorOr<u32> TagTree::read_value(u32 x, u32 y, Function<ErrorOr<bool>()> const&
+            size_t code_block_y = code_block_index / num_code_blocks_wide;
+
+            auto code_block_rect = IntRect { { code_block_x * (1 << xcb_prime), code_block_y * (1 << ycb_prime) }, { 1 << xcb_prime, 1 << ycb_prime } };
+            code_block_rect.set_location(code_block_rect.location() + rect_covered_by_codeblocks.location());
+
+            // B.7 Division of the sub-bands into code-blocks
+            // "NOTE – Code-blocks in the partition may extend beyond the boundaries of the sub-band coefficients. When this happens, only the
+            //  coefficients lying within the sub-band are coded using the method described in Annex D. The first stripe coded using this method
+            //  corresponds to the first four rows of sub-band coefficients in the code-block or to as many such rows as are present."
+            current_block.rect = code_block_rect.intersected(sub_band.rect);
+        }
+
+        if (!precinct.code_blocks.is_empty()) {
+            precinct.code_block_inclusion_tree = TRY(JPEG2000::TagTree::create(num_code_blocks_wide, num_code_blocks_high));
+            precinct.p_tree = TRY(JPEG2000::TagTree::create(num_code_blocks_wide, num_code_blocks_high));
+        }
+
+        return precinct;
+    };
+
+    auto make_sub_band = [&](TileData const& tile, int component_index, DecodedSubBand& sub_band, JPEG2000::SubBand sub_band_type, int r) -> ErrorOr<void> {
+        auto const& coding_parameters = context.coding_style_parameters_for_component(tile, component_index);
+        auto N_L = coding_parameters.number_of_decomposition_levels;
+
+        // Table F.1 – Decomposition level nb for sub-band b
+        // Note: The spec suggests that this ends with n_b = 1, but if N_L is 0, we have 0LL and nothing else.
+        auto n_b = [N_L](int r) { return r == 0 ? N_L : (N_L + 1 - r); };
+
+        sub_band.rect = context.siz.reference_grid_coordinates_for_sub_band(tile.rect, component_index, n_b(r), sub_band_type);
+
+        dbgln_if(JPEG2000_DEBUG, "Sub-band rect: {}", sub_band.rect);
+
+        // Compute tile size at resolution level r.
+        auto ll_rect = context.siz.reference_grid_coordinates_for_ll_band(tile.rect, component_index, r, N_L);
+
+        // B.6
+        // (B-16)
+        int num_precincts_wide = 0;
+        int num_precincts_high = 0;
+        int PPx = coding_parameters.precinct_sizes[r].PPx;
+        int PPy = coding_parameters.precinct_sizes[r].PPy;
+
+        if (ll_rect.width() != 0)
+            num_precincts_wide = ceil_div(ll_rect.right(), 1 << PPx) - (ll_rect.left() / (1 << PPx));
+        if (ll_rect.height() != 0)
+            num_precincts_high = ceil_div(ll_rect.bottom(), 1 << PPy) - (ll_rect.top() / (1 << PPy));
+
+        sub_band.num_precincts_wide = num_precincts_wide;
+        sub_band.num_precincts_high = num_precincts_high;
+
+        auto precinct_origin = IntPoint { ll_rect.x() & ~((1 << PPx) - 1), ll_rect.y() & ~((1 << PPy) - 1) };
```

Comment:
Nit: Brain tells me that the mask boils down to `(~0)<< PP[xy]`, not sure which version is easier to understand though

Reply:
My 2 cents: I find Nico's version easier to comprehend (but this is also the way I usually do such masks)

Reply:
Doesn't matter super much imho, so let's keep as-is for now :)